src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -216,6 +216,8 @@ internal void ReleaseRef()
         private DateTime _exitTime;
         /// <summary>A lazily-initialized event set when the process exits.</summary>
         private ManualResetEvent? _exitedEvent;
+        /// <summary>Whether allow reaping by other thread.</summary>
+        private static readonly bool s_allowReapingByOthers = Environment.GetEnvironmentVariable("DOTNET_AllowReapingByOthers") == "1";
```

Comment:
Do we really need this behavior to be configurable? What's the downside of allowing this by default?

Reply:
I don't see any downside. After all, it is a problem if a child is not reaped and keep consuming kernel resource.
But I'd like to honor the original author's intention that I might be not aware of.

Reply:
The downside is that we're working around an issue in a 3rd party library which causes `Process.ExitCode` to no longer be correct.

Reply:
What is the default behavior of other frameworks (Python, Java, ...)? Do they also fail fast in this situation, or do they handle it more gracefully?

I think it may be better to throw exception from `Process.ExitCode` with an error message that describes the problem. No opt-in necessary.

Reply:
This is how the issue occurs:

An app starts a child using `Process`.
And it also uses some native library which also starts a child.
In an attempt to reap that child, the native library reaps the managed process instead.
Then when .NET tries to reap it, it finds the process is already gone and does a `FailFast`.

The proper fix is to change the native library so it only reaps the child processes it started.
https://github.com/WiringPi/WiringPi/pull/101/files was the fix for https://github.com/dotnet/runtime/issues/33297.

@HJLeee would like to work around the bug in the native library without it affecting the overall app. Throwing from `Process.ExitCode` doesn't do that.

Reply:
Returning bogus values from `Process.ExitCode` is not going to do that either.

Reply:
Agree.

I think the `FailFast` is the appropriate when we detect the corrupted state.

I'm not in favor of introducing this workaround. The fix should happen in the native library.

Reply:
https://github.com/python/cpython/blob/13c10bfb777483c7b02877aab029345a056b809c/Lib/asyncio/unix_events.py#L1075-L1084
I presume this is how python handles reaping by others.
It sets 255 as return code.

When the return code of from native is not possible to convert, it returns status object.
https://github.com/python/cpython/blob/13c10bfb777483c7b02877aab029345a056b809c/Lib/asyncio/unix_events.py#L50-L54

Reply:
https://docs.python.org/3/library/asyncio-policy.html#process-watchers
It might help making decision for dotnet behavior.

Reply:
https://android.googlesource.com/platform/libcore/+/refs/heads/master/ojluni/src/main/java/java/lang/UNIXProcess.java#172
https://android.googlesource.com/platform/libcore/+/refs/heads/master/ojluni/src/main/native/UNIXProcess_md.c#341

It seems Android sets 0 as exit code for reaped process.
The code itself from openJDK.

Reply:
@HJLeee considering your use-case, what do you think is the appropriate exit code?

I think you are shifting the problem to some other place, hoping to reduce its impact.
Probably you don't control all `ExitCode` checking, so this introduces some undefined behavior into the app?

Also, are you really in a case where you can't fix, or ask your vendor to fix the issue in the native library?

@jkotas if this gets added, I think it should be opt-in. If this were my app, I'd like to know a native library is introducing a problem rather than .NET trying to cover it up.

There is no functional relationship between the .NET child processes that are effected and those started by the native library. That already makes the issue hard to detect and diagnose.

Reply:
If `ProcessWaitState` has exited but exitcode is null, it can be considered as reaped by others.
When `Process.ExitCode` called, it can throw an exception.

This is another option.

Reply:
It is elementary forked git. Currently the core of it is rather not maintained at all.

> Throwing an exception is similar to the FailFast because it is unexpected. It will cause the underlying issue to remain hidden in more cases, which is both good and bad.

It is unexpected but exception opens an option to the user as it can be documented and handled gracefully.
This particular underlying issue is not also concerned in other runtimes. So I think it is ok to align with them.
I agree with you that it can be both good and bad.

Reply:
> It is elementary forked git. Currently the core of it is rather not maintained at all.

The issue is http://git.enlightenment.org/enlightenment/efl/src/commit/919e646ad3253874f8979e153ecde11c2630b2e4/src/lib/ecore/ecore_signal.c#L418

It should change similar to what we have here:

https://github.com/dotnet/runtime/blob/702c1ea42dcd7c7d0a1ff139bda13b68033dc5ff/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs#L595-L597

Reply:
> It should change similar to what we have here:

I didn't spend much time but I was wondering this code is similar to ecore_signal strategy.
https://github.com/dotnet/runtime/blob/4c6b53b04a4e53f3cd6b19811c64ab608032265c/src/native/libs/System.Native/pal_signal.c#L370

Reply:
> We are generally not taking this stance. For example, we had requests to enable masking of crashes or memory management bugs in native libraries that we have refused.

Thank you for sharing general stance for handling outside bugs. Let me create an issue before making PR next time for discussion.

Reply:
The case for dart runtime.

https://github.com/dart-lang/sdk/blob/3634234ca000428fbb912cb42301ae8f2f143746/runtime/bin/process_linux.cc#L203-L235
It calls `wait()` to get exitcode which reaps any child and no guarantee for reaping its own child only.